### PR TITLE
feat: add k8s cronjob for rabbitmq backup, and re-factor repo [migration]

### PR DESCRIPTION
**Pull request by artsyjian**
[PLATFORM-4953]

- Add `src/rabbitmq_export`. It exports broker definitions from staging/production RabbitMQ. Remove old [Bash script](https://github.com/artsy/opstools/tree/main/src/rabbitmq). Add a k8s cronjob for each environment.
- The old Bash script only kept one backup - each backup overwrites the previous. The new script does not overwrite, therefore pruning is required. [src/kubernetes_backup](https://github.com/artsy/opstools/blob/main/src/kubernetes_backup/prune.py) already has pruning code which can be re-used. Therefore:
  - Extract `src/kubernetes_backup/prune.py` into an independent script `src/s3_prune_backups` which prunes for both k8s and RabbitMQ backups. Add k8s cronjobs.
  - Rename `src/kubernetes_backup` to `src/kubernetes_export` because now it does export only.
- About the k8s-related scripts (`src/kubernetes_...`), they must be aware of whether they are running inside the target k8s cluster. Currently they infer that from two vars: K8S_CLUSTER, KUBECTL_CONTEXT var. K8S_CLUSTER being set means in cluster, KUBECTL_CONTEXT means outside cluster. I find this approach too confusing. Re-factor as follows:
  - Add an `in_cluster` cmdline boolean option. `True` means script is running inside cluster. `False` otherwise.
  - Add a required `artsy_env` cmdline arg with two possible values: `staging`, `production`.
- Simplify `src/kubernetes_cleanup_jobs`. It supports cleaning all pods, pods by name, pods by completed status. We only clean up all jobs and I don't foresee us using the other two, so delete them.
- Re-factor `src/kubernetes_cleanup_pods`. It supports cleaning up pods by name or by completed status, in separate functions. However, the two criteria are not mutually exclusive. For example, I may be interested in pods that are named `foo` _and_  are completed. Replace the functions with a single function that provides the flexibility.
- Refactor `Kctl#_run`. Currently it raises exception when the `kubectl` commands returns non-zero. That is not always desirable for commands such as `kubectl delete pod ...` for which we may want to tolerate 'pod not found' (possibly it exited on its own, or was deleted by another job or by a human). Add an `expect_success` flag to give this clients option on whether exception should be raised.
- Lots of miscellaneous changes on: naming, styling, consistency, moving code into `src/lib`, simplification, etc.

Migration:
---
### Pre-merge
- [x] Update shared env.
- [x] Delete Staging cronjobs (due to changes in job names, cmdline args, ...)
  ```
  kubectl --context staging delete cronjob opstools-cleanup-completed-pods
  kubectl --context staging delete cronjob opstools-cleanup-hokusai-run-pods
  kubectl --context staging delete cronjob opstools-cleanup-jobs
  kubectl --context staging delete cronjob opstools-cleanup-review-apps
  kubectl --context staging delete cronjob opstools-k8s-create-backup
  kubectl --context staging delete cronjob opstools-k8s-prune-backup
  kubectl --context staging delete cronjob opstools-terraform-drift-detection
  ```
- [x]  Update Staging env.
  ```
  # delete obsolete vars
  hokusai staging env unset K8S_BACKUP_KEEPN
  hokusai staging env unset K8S_CLUSTER
  hokusai staging env unset S3_BUCKET
  hokusai staging env unset S3_PREFIX

  # set vars
  hokusai staging env set K8S_BACKUP_S3_BUCKET=<FILTERED>
  hokusai staging env set K8S_BACKUP_S3_PREFIX=<FILTERED>
  hokusai staging env set RABBITMQ_BACKUP_S3_BUCKET=<FILTERED>
  hokusai staging env set RABBITMQ_BACKUP_S3_PREFIX=<FILTERED>
  hokusai staging env set BACKUP_S3_BUCKET=<FILTERED>
  hokusai staging env set BACKUP_S3_PREFIX=<FILTERED>
  ```
- [x] Move staging k8s backup files to `s3://<FILTERED>/<FILTERED>/archive/staging/`. This is because of change in file naming - [timezone info will be added to filenames](https://github.com/artsy/opstools/compare/main...artsyjian/rabbit#diff-caa5992b80ab21c80ef723d3acc57cd7745008cb08dd18ecd19b6cc6e258b1e2R37).

### Staging QA
Mostly running the jobs and check output. Jobs are safe to run.
```
kubectl --context staging create job opstools-cleanup-completed-pods-test --from=cronjob/opstools-cleanup-completed-pods
kubectl --context staging get pod | grep opstools-cleanup-completed-pods-test
kubectl --context staging logs -f <pod name>

kubectl --context staging create job opstools-cleanup-hokusai-run-pods-test --from=cronjob/opstools-cleanup-hokusai-run-pods
kubectl --context staging get pod | grep opstools-cleanup-hokusai-run-pods-test
kubectl --context staging logs -f <pod name>

kubectl --context staging create job opstools-cleanup-jobs-test --from=cronjob/opstools-cleanup-jobs
kubectl --context staging get pod | grep opstools-cleanup-jobs-test
kubectl --context staging logs -f <pod name>

kubectl --context staging create job opstools-cleanup-review-apps-test --from=cronjob/opstools-cleanup-review-apps
kubectl --context staging get pod | grep opstools-cleanup-review-apps
kubectl --context staging logs -f <pod name>

kubectl --context staging create job opstools-k8s-export-test --from=cronjob/opstools-k8s-export
kubectl --context staging get pod | grep opstools-k8s-export-test
kubectl --context staging logs -f <pod name>
...look for an S3 location in logs output, and confirm the location has a .tar.gz file
...confirm the filename has a 'Z' which denotes UTC timezone

kubectl --context staging create job opstools-rabbitmq-export-test --from=cronjob/opstools-rabbitmq-export
kubectl --context staging get pod | grep opstools-rabbitmq-export-test
kubectl --context staging logs -f <pod name>
...look for the S3 location in logs and confirm it has a .json file

kubectl --context staging create job opstools-s3-prune-backups-k8s-test --from=cronjob/opstools-s3-prune-backups-k8s
kubectl --context staging get pod | grep opstools-s3-prune-backups-k8s-test
kubectl --context staging logs -f <pod name>

kubectl --context staging create job opstools-s3-prune-backups-rabbitmq-test --from=cronjob/opstools-s3-prune-backups-rabbitmq
kubectl --context staging get pod | grep opstools-s3-prune-backups-rabbitmq-test
kubectl --context staging logs -f <pod name>

kubectl --context staging create job opstools-terraform-drift-detection-test --from=cronjob/opstools-terraform-drift-detection
kubectl --context staging get pod | grep opstools-terraform-drift-detection-test
kubectl --context staging logs -f <pod name>
```
### Pre-merge-Deploy-PR
- [ ] Delete Production cronjobs (due to changes in job names, cmdline args, ...)
  ```
  kubectl --context staging delete cronjob opstools-cleanup-completed-pods
  kubectl --context staging delete cronjob opstools-cleanup-hokusai-run-pods
  kubectl --context staging delete cronjob opstools-cleanup-jobs
  kubectl --context staging delete cronjob opstools-ecr-check-repos-for-terraform
  kubectl --context staging delete cronjob opstools-k8s-create-backup
  kubectl --context staging delete cronjob opstools-k8s-prune-backup
  kubectl --context staging delete cronjob opstools-terraform-drift-detection
  ```
- [ ]  Update Production env.
  ```
  # delete obsolete vars
  hokusai production env unset K8S_BACKUP_KEEPN
  hokusai production env unset K8S_CLUSTER
  hokusai production env unset S3_BUCKET
  hokusai production env unset S3_PREFIX
  
  # set vars
  hokusai production env set K8S_BACKUP_S3_BUCKET=<FILTERED>
  hokusai production env set K8S_BACKUP_S3_PREFIX=<FILTERED>
  hokusai production env set RABBITMQ_BACKUP_S3_BUCKET=<FILTERED>
  hokusai production env set RABBITMQ_BACKUP_S3_PREFIX=<FILTERED>
  hokusai production env set BACKUP_S3_BUCKET=<FILTERED>
  hokusai production env set BACKUP_S3_PREFIX=<FILTERED>
  ```
- [ ] Delete RabbitMQ Jenkins jobs.
- [ ] Move production k8s backup files to `s3://<FILTERED>/<FILTERED>/archive/staging/`. This is because of change in file naming - [timezone info will be added to filenames](https://github.com/artsy/opstools/compare/main...artsyjian/rabbit#diff-caa5992b80ab21c80ef723d3acc57cd7745008cb08dd18ecd19b6cc6e258b1e2R37).

### Prod QA
Mostly running the jobs and check output. Jobs are safe to run.
```
kubectl --context production create job opstools-cleanup-completed-pods-test --from=cronjob/opstools-cleanup-completed-pods
kubectl --context production get pod | grep opstools-cleanup-completed-pods-test
kubectl --context production logs -f <pod name>

kubectl --context production create job opstools-cleanup-hokusai-run-pods-test --from=cronjob/opstools-cleanup-hokusai-run-pods
kubectl --context production get pod | grep opstools-cleanup-hokusai-run-pods-test
kubectl --context production logs -f <pod name>

kubectl --context production create job opstools-cleanup-jobs-test --from=cronjob/opstools-cleanup-jobs
kubectl --context production get pod | grep opstools-cleanup-jobs-test
kubectl --context production logs -f <pod name>

kubectl --context production create job opstools-ecr-check-repos-for-terraform-test --from=cronjob/opstools-ecr-check-repos-for-terraform
kubectl --context production get pod | opstools-ecr-check-repos-for-terraform-test
kubectl --context production logs -f <pod name>

kubectl --context production create job opstools-k8s-export-test --from=cronjob/opstools-k8s-export
kubectl --context production get pod | grep opstools-k8s-export-test
kubectl --context production logs -f <pod name>
...look for an S3 location in logs output, and confirm the location has a .tar.gz file
...confirm the filename has a 'Z' which denotes UTC timezone

kubectl --context production create job opstools-rabbitmq-export-test --from=cronjob/opstools-rabbitmq-export
kubectl --context production get pod | grep opstools-rabbitmq-export-test
kubectl --context production logs -f <pod name>
...look for the S3 location in logs and confirm it has a .json file

kubectl --context production create job opstools-s3-prune-backups-k8s-test --from=cronjob/opstools-s3-prune-backups-k8s
kubectl --context production get pod | grep opstools-s3-prune-backups-k8s-test
kubectl --context production logs -f <pod name>

kubectl --context production create job opstools-s3-prune-backups-rabbitmq-test --from=cronjob/opstools-s3-prune-backups-rabbitmq
kubectl --context production get pod | grep opstools-s3-prune-backups-rabbitmq-test
kubectl --context production logs -f <pod name>

kubectl --context production create job opstools-terraform-drift-detection-test --from=cronjob/opstools-terraform-drift-detection
kubectl --context production get pod | grep opstools-terraform-drift-detection-test
kubectl --context production logs -f <pod name>
```

[PLATFORM-4953]: https://artsyproduct.atlassian.net/browse/PLATFORM-4953?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/ecr_check_repos_for_terraform/check.py [hunk 1]
```diff
@@ -15,4 +15,4 @@
     logging.info(f"Repositories found: {repos}")
     exit(1)
   else:
-    logging.info("None found.")
+    logging.info("None repositories found.")
```
Comment (mc-jones) [inline src/ecr_check_repos_for_terraform/check.py:18]: Minor: suggest `No repositories found.`

Comment (artsyjian) [reply src/ecr_check_repos_for_terraform/check.py]: good catch! thanks.

File: src/kubernetes_cleanup_pods/test/test_pods.py [hunk 2]
```diff
@@ -27,6 +28,11 @@ def it_calls_pods_obj_delete(mocker, pods_obj):
         mocker.call('pod1'),
         mocker.call('pod2')
       ])
+    def it_refuses_to_delete_more_than_30_pods(mocker, pods_obj):
```
Comment (mc-jones) [inline src/kubernetes_cleanup_pods/test/test_pods.py:31]: Curious if there is a specific reason for this or if it is just a reasonable guard?

Comment (artsyjian) [reply src/kubernetes_cleanup_pods/test/test_pods.py:31]: A guard - to prevent accidentally deleting all pods in the cluster.

**Comment by mc-jones** [comment]
Whew 😅 , what an epic PR! All the changes seem reasonable to me but it may make sense to get another review given the size of this PR and my inexpertise with Python. 

My main comment would just be that given the size and particular sensitivity of the operations involved, it would be helpful to understand the testing plan a bit more (though test coverage looks robust). It seems like all jobs/operations are fully running in staging, so we should be able to test there for a period. Is this the plan, or did you have something else in mind?

**Comment by artsyjian** [comment]
> My main comment would just be that given the size and particular sensitivity of the operations involved, it would be helpful to understand the testing plan a bit more (though test coverage looks robust).

Agreed. I'll hash out the test steps and update PR.

**Comment by artsyjian** [comment]
Added Staging/Prod QA steps.

**Review by mc-jones — APPROVED** [review]

**Comment by artsyjian** [comment]
migration applied. minor problem encountered and fixed in https://github.com/artsy/opstools/pull/37

**Pull request merged**